processor/transformprocessor/internal/metrics/func_extract_sum_metric.go

```diff
@@ -0,0 +1,117 @@
+}
+
+func newExtractSumMetricFactory() ottl.Factory[ottldatapoint.TransformContext] {
+	return ottl.NewFactory("extract_sum_metric", &extractSumMetricArguments{}, createExtractSumMetricFunction)
+}
+
+func createExtractSumMetricFunction(_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[ottldatapoint.TransformContext], error) {
+	args, ok := oArgs.(*extractSumMetricArguments)
+
+	if !ok {
+		return nil, fmt.Errorf("extractSumMetricFactory args must be of type *extractSumMetricArguments")
+	}
+
+	return extractSumMetric(args.StringAggTemp, args.Monotonic)
+}
+
+// this interface helps unify the logic for extracting data from different histogram types
+// all supported metric types' datapoints implement it
+type SumCountDataPoint interface {
+	Attributes() pcommon.Map
+	Sum() float64
+	Count() uint64
+	StartTimestamp() pcommon.Timestamp
+	Timestamp() pcommon.Timestamp
+}
+
+func extractSumMetric(stringAggTemp string, monotonic bool) (ottl.ExprFunc[ottldatapoint.TransformContext], error) {
+	var aggTemp pmetric.AggregationTemporality
+	switch stringAggTemp {
+	case "delta":
+		aggTemp = pmetric.AggregationTemporalityDelta
+	case "cumulative":
+		aggTemp = pmetric.AggregationTemporalityCumulative
+	default:
+		return nil, fmt.Errorf("unknown aggregation temporality: %s", stringAggTemp)
+	}
+	return func(_ context.Context, tCtx ottldatapoint.TransformContext) (interface{}, error) {
+		metric := tCtx.GetMetric()
+
+		switch metric.Type() {
+		case pmetric.MetricTypeHistogram, pmetric.MetricTypeExponentialHistogram, pmetric.MetricTypeSummary:
+		default:
+			return nil, nil
+		}
+
+		// We're in a datapoint context, but we don't want to create a new Metric for each datapoint
+		// so we first check if it doesn't already exist
+		sumMetric := getOrCreateSumMetric(
+			tCtx.GetMetrics(),
+			metric.Name()+"_sum",
+			metric.Description(),
+			metric.Unit(),
+			aggTemp,
+			monotonic,
+		)
+
+		histogramDatapoint := tCtx.GetDataPoint().(SumCountDataPoint)
+		sumDp := sumMetric.Sum().DataPoints().AppendEmpty()
+		histogramDatapoint.Attributes().CopyTo(sumDp.Attributes())
+		sumDp.SetDoubleValue(histogramDatapoint.Sum())
+		sumDp.SetStartTimestamp(histogramDatapoint.StartTimestamp())
+		sumDp.SetTimestamp(histogramDatapoint.Timestamp())
+
+		return nil, nil
+	}, nil
+}
+
+// getOrCreateSumMetric looks for the sum metric in the metric slice, and creates one if it can't find it
+func getOrCreateSumMetric(
+	metrics pmetric.MetricSlice,
+	name string,
+	description string,
+	unit string,
+	aggregationTemporality pmetric.AggregationTemporality,
+	isMonotonic bool,
+) pmetric.Metric {
+	// try to find the metric in the slice
+	for i := 0; i < metrics.Len(); i++ {
+		metric := metrics.At(i)
+		found := (metric.Name() == name &&
+			metric.Description() == description &&
+			metric.Unit() == unit &&
+			metric.Type() == pmetric.MetricTypeSum &&
+			metric.Sum().IsMonotonic() == isMonotonic)
+		if found {
+			return metric
+		}
+	}
```

Comment:
Is there something we can add to the datapoint context's `TransformContext` to make this lookup easier?  Maybe each TransformContext could keep a map of the associated metrics that could be quickly accessed?  It would be nice to avoid looping over each metric for every datapoint.

Reply:
I considered putting it in cache, but I wasn't sure I could guarantee no collisions with user-provided keys. Or do you mean adding a new field to the `TransformContext` struct?

Reply:
A new field on the TransformContext struct.  I am thinking the struct could have a map and then a new function on the interface like `GetMetricMap` that would return it.  We'd have to think about how to ensure the map says up to date when the metric slice is appended to though and I think we'd need it to be read only.

It would probably be worth it to do some performance benchmarks on the loop solution and then compare to a map solution and see how much we actually save.  The complexity might not be worth it and right now it seems like we'd be in this loop a lot, which will be problematic for big metric batches.

Reply:
This metric is always appended to the end of the slice, so if I iterate over it in reverse, I should find it immediately if it exists, and only loop through the whole thing if it doesn't. I can also only check the last metric in the slice and create if that one doesn't fit. Do we give any guarantees in OTTL about the order in which datapoints are processed? I can see the implementation processes all datapoints in a single metric in order, but I don't want to create potential performance regressions in the future by making assumptions about implementation details here.

I'll do a simple benchmark and see how they compare.

Reply:
> Do we give any guarantees in OTTL about the order in which datapoints are processed?

No.  Everything happens to be in order because that is how pdata handles loops, but the transformprocessor/OTTL doesn't guarantee that that are handled in the order the collector received them.

Reply:
This conversation makes me want to try out the metric context approach instead, to be honest. The solution I went with is looking more magical the more I think about it.

Reply:
Another option is to create lots of metrics with single datapoints.

Reply:
I've changed to that implementation in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24368/commits/96944dc79acb3e8a56da7952ea3914923d6c902b and https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24368/commits/7281e3a0b69485d6c976a577667406911bc09397. If we decide to go this way, this should probably go into a separate PR, but I do like this more than the datapoint context version. WDYT?

Reply:
I like the look of that and I don't see any different between ottldatapoint have access to the metricslice and ottlmetric having access to the metric slice; either want the context has access to the slice while "inside" it.